res/controllers/Pioneer-DDJ-400-script.js

```diff
@@ -0,0 +1,668 @@
+        engine.setValue(group, 'cue_gotoandstop', 1);
+    } else { // load selected track
+        engine.setValue(group, 'LoadSelectedTrack', 1);
+    }
+    // TODO: while playing a sample blink playing PAD?
+};
+
+// nschloe: This section handles the beat jump mode. It's not working yet, so in the XML
+// we simply use the intrinsic beatjump_*_{forward,backward} functions for now.
+//
+// PioneerDDJ400.beatjumpPadPressed = function(_channel, control, value, status, group) {
+//     'use strict';
+//     if (value === 0) {
+//         midi.sendShortMsg(status, control, 0); // turn off LED
+//         return; // ignore release
+//     }
+//
+//     var newVal = this.beatjumpPad[control & 0xf] * this.beatjumpMultiplier;
+//
+//     engine.setValue(group, 'beatjump', newVal);
+//     midi.sendShortMsg(status, control, 0x7f); // turn LED back on
+// };
+//
+// PioneerDDJ400.beatjumpShiftByOne = function(_channel, control, value, status) {
+//     'use strict';
+//     if (value === 0) {
+//         midi.sendShortMsg(status, control, 0); // turn off LED
+//         return; // ignore release
+//     }
+//
+//     var direction = status <= 0x98 ? -1 : 1;
+//
+//     if ( direction === -1 && this.beatjumpMultiplier <= 1) {
+//         direction = 0;
+//     }
+//
+//     this.beatjumpMultiplier += direction;
+//     midi.sendShortMsg(status, control, 0x7f); // turn LED back on
+// };
+
+PioneerDDJ400.shiftPressed = function(channel, _control, value) {
+    'use strict';
+    this.shiftState[channel] = value;
+};
+
+PioneerDDJ400.waveFormRotate = function(_channel, _control, value) {
+    'use strict';
+    // select the Waveform to zoom left shift = deck1, right shift = deck2
+    var deckNum = this.shiftState[0] > 0 ? 1 : 2;
+    var oldVal = engine.getValue('[Channel'+deckNum+']', 'waveform_zoom');
+    var newVal = oldVal + (value > 0x64 ? 1 : -1);
+
+    engine.setValue('[Channel'+deckNum+']', 'waveform_zoom', newVal);
+};
+
+PioneerDDJ400.loopin4beatPressed = function(channel, _control, value, _status, group) {
+    'use strict';
+    var loopEnabled = engine.getValue(group, 'loop_enabled');
+    this.loopin4beat[channel] = (value > 0);
+
+    if (loopEnabled === 0 && value > 0) {
+        engine.setValue(group, 'loop_in', 1);
+    }
+};
+
+PioneerDDJ400.loopin4beatPressedLong = function(_channel, _control, value, _status, group) {
+    'use strict';
+    var loopEnabled = engine.getValue(group, 'loop_enabled');
+    if (!loopEnabled && value > 0) {
+        engine.setValue(group, 'beatloop_4_activate', 1);
+    }
+};
+
+PioneerDDJ400.loopoutPressed = function(channel, _control, value, _status, group) {
+    'use strict';
+    var loopEnabled = engine.getValue(group, 'loop_enabled');
+    this.loopout[channel] = (value > 0);
+
+    if (loopEnabled === 0 && value > 0) {
+        engine.setValue(group, 'loop_out', 1);
+    }
+};
+
+// START BEAT FX
+
+PioneerDDJ400.numFxSlots = 3;
+
+Object.defineProperty(PioneerDDJ400, 'selectedFxSlot', {
```

Comment:
The Components EffectUnit is inappropriate for this controller. I proposed how to map the effects on the [old PR](https://github.com/mixxxdj/mixxx/pull/2385#issuecomment-563053930).

Reply:
To make things easier, here are @Be-ing's comments:
> I have a better idea for the effects. Instead of scrolling forward/back for focusing effects, have a dedicated button to focus each effect:
>
> Level/Depth knob: metaknob of focused effect; if no effect is focused, controls the mix knob for effect unit 1
> Beat <: focus effect 1
> Beat >: focus effect 2
> FX select button: focus effect 3
> On/Off button: enable/disable focused effect
> 1/2/Master switch: assign effect unit 1 to deck 1/2/master as labeled
> Unfocus by pressing the button for the focused effect again (so if effect 1 is focused, press Beat < to unfocus and make the knob control the mix knob of the unit).

Unfortunately, the implementation of these suggestions is way over my head. I'll be happy to merge any related PRs. I also wouldn't mind removing the effects section entirely if it's not up to the standards.

Reply:
I am reluctant to merge this without a more useful way to map the effects section. @dj3730 would you be able to implement this?

Reply:
@Be-ing Unfortunately I'm going to be tied up for the next couple weeks, I have a gig coming up next weekend that I'm prepping for, and it''l take me a few days after that to get back to hacking projects.

Having said that, your proposed changes don't see too complex.  A few questions, though.

The current script "hard-wires" Effect Unit 1 to Deck 1, and Effect Unit 2 to Deck 2, in the init code.  Effect Unit 3 can then be switched back and forth between the two decks, or enabled for both decks, using the slider switch.  Under your proposal, would we drop support for effect units 2 and 3 entirely and just focus on controlling Effect Unit 1?

Next. in the current script, pressing Shift_EffectOnOff disables all effects in the effects bank - would we want to keep that?

Finally, by using the FX Select button to focus effect 3 in Effect Unit 1, we loose the ability to change the selected effect.  Not sure if this is the worst thing though, as I find it cumbersome to move through the list of possible effects this way and generally just use the mouse for that, but it bears considering.

So it'll be until around the 20th of this month before I can really look into this.  If that's an acceptable delay, I'd be willing to take a whack at it.

Reply:
>  Under your proposal, would we drop support for effect units 2 and 3 entirely and just focus on controlling Effect Unit 1?

Yes

> Next. in the current script, pressing Shift_EffectOnOff disables all effects in the effects bank - would we want to keep that?

Considering there is only one button and you'd have to switch focus between effects to enable/disable each of them, this seems like a good idea to me.

> Finally, by using the FX Select button to focus effect 3 in Effect Unit 1, we loose the ability to change the selected effect. Not sure if this is the worst thing though, as I find it cumbersome to move through the list of possible effects this way and generally just use the mouse for that, but it bears considering.

I agree, selecting effects from controllers is more trouble than it's worth.

> So it'll be until around the 20th of this month before I can really look into this. If that's an acceptable delay, I'd be willing to take a whack at it.

That's okay, thanks for picking this up.